tests/validators/test_dataclasses.py

```diff
@@ -714,7 +714,7 @@ def test_dataclass_validate_assignment():
     ]
 
     # wrong arguments
-    with pytest.raises(TypeError, match='self_instance should not be None on dataclass validate_assignment'):
+    with pytest.raises(TypeError, match="'field_a' is not a model instance"):
```

Comment:
I think this error message is better, a hypothetical user of pydantic-core that doesn't know the internals would not know what `self_instance` means.